akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowLimitSpec.scala

```diff
@@ -0,0 +1,54 @@
+package akka.stream.scaladsl
+
+import akka.stream.{ StreamLimitReachedException, ActorMaterializer, ActorMaterializerSettings }
+import akka.stream.testkit.AkkaSpec
+import scala.concurrent.Await
+import scala.concurrent.duration._
+
+class FlowLimitSpec extends AkkaSpec {
+
+  val settings = ActorMaterializerSettings(system)
+    .withInputBuffer(initialSize = 2, maxSize = 16)
+
+  implicit val mat = ActorMaterializer(settings)
+
+  "Limit" must {
+    "produce output that is identical to the input when n = input.length" in {
+      val input = (1 to 6)
+      val n = input.length
+      val future = Source(input).limit(n).grouped(Integer.MAX_VALUE).runWith(Sink.head)
+      val result = Await.result(future, 300.millis)
+      result should be(input.toSeq)
+    }
+
+    "produce output that is identical to the input when n > input.length" in {
+      val input = (1 to 6)
+      val n = input.length + 2 // n > input.length
+      val future = Source(input).limit(n).grouped(Integer.MAX_VALUE).runWith(Sink.head)
+      val result = Await.result(future, 300.millis)
+      result should be(input.toSeq)
+    }
+
+    "produce n messages before throwing a StreamLimitReachedException when n < input.size" in {
+      // TODO: check if it actually produces n messages
```

Comment:
doing this means using `recover` after the `limit` stage so that `grouped` actually gets a chance to emit things